File: polyglot/ruby/file_carving.rb
# polyglot/ruby/file_carving.rb

require 'fileutils'

module FileCarving
  class Carver
    attr_reader :image_path, :output_dir, :signatures

    def initialize(image_path, output_dir, signatures)
      @image_path = image_path
      @output_dir = output_dir
      @signatures = signatures
    end

    def carve!
      FileUtils.mkdir_p(@output_dir) unless File.directory?(@output_dir)

      File.open(@image_path, 'rb') do |image|
        buffer = ''
        current_file = nil

        image.each_byte do |byte|
          buffer += [byte].pack('C')

          # Check if buffer matches any signature
          @signatures.each do |signature|
            if buffer.end_with?(signature)
              # Start of file found, prepare to carve
              current_file = {
                name: generate_filename(signature),
                content: buffer,
                offset: image.pos - signature.size
              }
              break
            end
          end

          # If a file was started, check for end of file
          if current_file
            @signatures.each do |signature|
              if buffer.end_with?(signature)
                # End of file found, write to disk
                write_file(current_file)
                current_file = nil
                break
              end
            end
          end
        end

        # Handle any remaining buffer that might be part of a file
        if current_file
          write_file(current_file)
        end
      end
    end

    private

    def generate_filename(signature)
      signature.gsub(/\W/, '_').downcase + '.bin'
    end

    def write_file(file_info)
      file_path = File.join(@output_dir, file_info[:name])
      File.open(file_path, 'wb') do |f|
        f.write(file_info[:content])
      end
      puts "Recovered file: #{file_path} (offset: #{file_info[:offset]})"
    end
  end

  def self.run(image_path, output_dir, signatures)
    Carver.new(image_path, output_dir, signatures).carve!
  end
end

# Demo entry point
if __FILE__ == $0
  # Example signatures (e.g., common file headers)
  signatures = [
    "\x50\x4b\x03\x04", # ZIP
    "\x49\x49\x2A\x00", # JPEG
    "\x42\x4D",         # BMP
    "\x89\x50\x4E\x47\x0D\x0A\x1A\x0A", # PNG
    "\x25\x50\x44\x46", # PDF
    "\x47\x49\x46\x38", # GIF
    "\x00\x00\x01\xB2", # MP3
    "\xFF\xD8\xFF",     # JPEG (alternative)
    "\x52\x49\x46\x46", # WAV
    "\x49\x49\x49\x49", # TIFF
  ]

  output_dir = 'carved_files'
  image_path = 'disk_image.raw' # Replace with your disk image path

  puts "Starting file carving..."
  FileCarving.run(image_path, output_dir, signatures)
  puts "File carving completed. Files saved to #{output_dir}."
end

File: polyglot/rust/file_carving.rs
use std::fs::File;
use std::io::{BufReader, BufRead};
use std::path::Path;

fn main() {
    // Example usage: carve files from a disk image using known file signatures
    let input_path = "disk_image.raw";
    let output_dir = "carved_files";

    // Create output directory if it doesn't exist
    std::fs::create_dir_all(output_dir).unwrap();

    // Open the disk image file
    let file = File::open(input_path).expect("Failed to open disk image");
    let reader = BufReader::new(file);

    // Known file signatures (magic numbers) for common file types
    let signatures = vec![
        ("PNG", b"\x89\x50\x4E\x47\x0D\x0A\x1A\x0A"),
        ("JPEG", b"\xFF\xD8\xFF"),
        ("ZIP", b"\x50\x4B\x03\x04"),
        ("PDF", b"%PDF-"),
        ("MP3", b"ID3"),
        ("GIF", b"GIF87a"),
        ("TIFF", b"II*\x00"),
        ("BMP", b"BM"),
        ("TXT", b"\x0D\x0A"),
    ];

    // Process each line of the disk image
    for (line_idx, line_result) in reader.lines().enumerate() {
        let line = line_result.expect("Failed to read line");
        let bytes: Vec<u8> = line.bytes().collect();

        // Check for file signatures in the current line
        for (name, sig) in &signatures {
            if bytes.windows(sig.len()).any(|window| window == sig) {
                let filename = format!("{}_{}", name, line_idx);
                let output_path = Path::new(output_dir).join(filename);

                // Write the line to the output file
                let mut file = File::create(&output_path).expect("Failed to create output file");
                file.write_all(line.as_bytes()).expect("Failed to write carved file");

                println!("Carved {} from line {}", filename, line_idx);
            }
        }
    }

    println!("File carving complete.");
}